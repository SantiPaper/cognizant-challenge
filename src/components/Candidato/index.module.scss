.containerCandidato{
    display: flex;
    margin-bottom: 1.5rem;
    padding: 1rem;
    background-color: white;
    justify-content: space-between;

    h3{
        font-size: 1.5rem;
    }
    p{
        font-size: 1.3rem;
    }
   
}

