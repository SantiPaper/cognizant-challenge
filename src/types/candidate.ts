export interface Candidate {
  id: string;
  name: string;
  comments: string;
  step: Step;
}

export type Step =
  | "Entrevista inicial"
  | "Entrevista técnica"
  | "Oferta"
  | "Asignación"
  | "Rechazo";
