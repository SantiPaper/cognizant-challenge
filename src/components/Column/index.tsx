import type {Candidate, Step} from "../../types/candidate";

import React, {FormEvent, FunctionComponent, useState} from "react";
import {v4} from "uuid";

import {Candidato} from "../Candidato";

import style from "./index.module.scss";

type Props = {
  candidatos: Candidate[];
  titulo: Step;
  showButton?: boolean;
  advance: (id: string, action: "avanzar" | "retroceder") => void;
  addCandidate: (candidato: Candidate) => void;
};

export const Column: FunctionComponent<Props> = ({
  candidatos,
  titulo,
  advance,
  showButton,
  addCandidate,
}) => {
  const [showForm, setShowForm] = useState(false);

  const crearCandidato = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const formData = new FormData(e.target);
    const newCandidate = Object.fromEntries(formData.entries());

    newCandidate.id = v4();
    newCandidate.step = "Entrevista inicial";

    addCandidate(newCandidate as unknown as Candidate);
  };

  return (
    <div className={style.column}>
      <h2>{titulo}</h2>

      {candidatos.length > 0 ? (
        candidatos.map((candidato) => (
          <Candidato key={candidato.id} advance={advance} candidato={candidato} />
        ))
      ) : (
        <p>No hay candidatos</p>
      )}

      {showButton && (
        <button
          onClick={() => {
            setShowForm(!showForm);
          }}
        >
          Agregar candidato
        </button>
      )}
      {showForm && (
        <form onSubmit={crearCandidato}>
          <label htmlFor="nombre">Nombre</label>
          <input id="nombre" name="name" type="text" />

          <label htmlFor="comentario">Algun comentario</label>
          <input id="comentario" name="comments" type="text" />

          <button>Dale loco</button>
        </form>
      )}
    </div>
  );
};
