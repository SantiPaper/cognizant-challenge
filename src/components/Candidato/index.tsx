import React, {FunctionComponent} from "react";

import {Candidate} from "../../types/candidate";

import style from "./index.module.scss";

type Props = {
  candidato: Candidate;
  advance: (id: string, action: "avanzar" | "retroceder") => void;
};

export const Candidato: FunctionComponent<Props> = ({candidato, advance}) => {
  return (
    <article className={style.containerCandidato}>
      <div>
        <h3>{candidato.name}</h3>
        <p>{candidato.comments}</p>
      </div>
      <div>
        {candidato.step !== "Entrevista inicial" && (
          <button
            aria-label="Retroceder proceso"
            onClick={() => advance(candidato.id, "retroceder")}
          >
            {"<"}
          </button>
        )}
        {candidato.step !== "Rechazo" && (
          <button aria-label="Avanzar proceso" onClick={() => advance(candidato.id, "avanzar")}>
            {">"}
          </button>
        )}
      </div>
    </article>
  );
};
