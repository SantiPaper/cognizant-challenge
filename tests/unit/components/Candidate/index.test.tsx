import React from "react";
import {render, screen} from "@testing-library/react";

import {Candidato} from "../../../../src/components/Candidato";
import {Candidate} from "../../../../src/types/candidate";

const candidate: Candidate = {
  name: "Santiago",
  comments: "Burro",
  id: "Santi",
  step: "Entrevista inicial",
};

const advance = vi.fn();

describe("Componente candidato", () => {
  // Snapshot
  test("Renderiza la informacion", () => {
    const {container} = render(<Candidato advance={advance} candidato={candidate} />);

    expect(container).toMatchSnapshot();
  });

  //OTRO TERMINO QUERY QSY
  test("Renderiza boton retroceder", () => {
    render(<Candidato advance={advance} candidato={{...candidate, step: "Entrevista técnica"}} />);

    const backButton = screen.getByRole("button", {name: "Retroceder proceso"});

    expect(backButton).toBeVisible();
  });

  test("No renderiza ultimo boton avanzar", () => {
    render(<Candidato advance={advance} candidato={{...candidate, step: "Rechazo"}} />);

    const advanceButton = screen.queryByRole("button", {name: "Avanzar proceso"});

    expect(advanceButton).not.toBeInTheDocument();
  });

  test("Advance recibe los parametros esperados", () => {
    render(<Candidato advance={advance} candidato={{...candidate, step: "Entrevista técnica"}} />);

    const advanceButton = screen.getByRole("button", {name: "Avanzar proceso"});
    const backButton = screen.getByRole("button", {name: "Retroceder proceso"});

    expect(advance).not.toHaveBeenCalled();

    advanceButton.click();
    expect(advance).toHaveBeenCalledWith(candidate.id, "avanzar");

    backButton.click();
    expect(advance).toHaveBeenCalledWith(candidate.id, "retroceder");
  });
});
