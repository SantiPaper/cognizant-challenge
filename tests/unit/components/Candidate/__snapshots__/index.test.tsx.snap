// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente candidato > Renderiza la informacion 1`] = `
<div>
  <article
    class="_containerCandidato_59e197"
  >
    <div>
      <h3>
        Santiago
      </h3>
      <p>
        Burro
      </p>
    </div>
    <div>
      <button
        aria-label="Avanzar proceso"
      >
        &gt;
      </button>
    </div>
  </article>
</div>
`;
